sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -75,6 +75,90 @@ trait ScalaReflection {
     case (other, _) => other
   }
 
+  /**
+   * Creates a converter function that will convert Scala objects to the specified catalyst type.
+   */
+  private[sql] def createCatalystConverter(dataType: DataType): (Any) => Any = {
+    def extractOption(item: Any) = item match {
+      case o: Some[_] => o.get
+      case other => other
+    }
+
+    dataType match {
+      // Check UDT first since UDTs can override other types
+      case udt: UserDefinedType[_] => (item) => {
+        if (item == None) null else udt.serialize(extractOption(item))
+      }
+
+      case arrayType: ArrayType => {
```

Comment:
Generally, please don't add braces in a `case` class, namely, prefer

```scala
foo match {
  case ... =>
    expr1
    expr2
}
```

over

```scala
foo match {
  case ... => {
    expr1
    expr2
  }
}
```

This rule applies to the whole PR.